README.md
# CALYPSO kit

This repository maintains toolkit for CALYPSO structure prediction software.

## How does this repo organized?

    /calypsokit
      |- analysis             # structure analysis
         |- validity.py       # structure validity
      |- calydb               # calypso structure database
         |- calydb.py
         |- query.py          # customized query methods

## CALYPSO Database

### Connect to the database

To connect it with pymongo, the database address, username, password are read from
the environment variables. You can either put them in you environment variables or
write a `.env` file:

```properties
MONGODB_ADDR=<ip>:<port>
MONGODB_USER=<username>
MONGODB_PWD=<password>
MONGODB_DATABASE=<database name>
```

### How does the DataBase organized?

The database is stored by MongoDB.

    /calydb            # database
      |- raw           # raw data collection
      |- uniq          # unique _id collection
      |- *             # other selected collection
      |- debugcol      # debug collection

### What are recorded in `rawcol`?


    "_id":                      ObjectId(***),    # bson ObjectId
    "material_id":              "caly-*****",     # str
    "source": {
        "name":         "calypso",      # calypso or materials project, etc.
        "index":        *****,          # int index in this source, or None
    },
    "elements":                 ['La', 'H'],     # list of elements str
    "nelements":                2,               # number of elements
    "elemcount":                [2, 20],         # list of int, count of each element
    "species":                  ['La', ...],     # list of str, species of each atom
    "formula":                  "La2H20",        # formula str, metal and alphabet order
    "reduced_formula":          "LaH10",         # formula str, metal and alphabet order
    "natoms":                   22,              # int
    "cell":                     3x3 ndarray,     # cell matrix in angstrom
    "cell_abc":                 [a, b, c],       # cell length, angstrom
    "cell_angles":              [α, β, γ],       # cell angles, angle
    "positions":                22x3 ndarray,    # atom cartesian coordinates in angstrom
    "scaled_positions":         22x3 ndarray,    # atom fractional coordinates in angstrom
    "forces":                   22x3 ndarray,    # force of each atom in eV/A
    "enthalpy":                 *.*,             # eV
    "enthalpy_per_atom":        *.*,             # eV/atom
    "volume":                   *.*,             # A^3
    "volume_per_atom":          *.*,             # A^3/atom
    "density":                  *.*,             # g/cm^3
    "clospack_volume":          *.*,             # float, A^3, close-pack volume of covalent radii
    "clospack_volume_per_atom": *.*,             # float, A^3
    "clospack_density":         *.*,             # float, g/cm^3
    "volume_rate":              *,      # float, volume/clospack_volume
    "pressure":                 *.*,    # float, GPa
    "pressure_range": {                         # each pressure is set to a bin
        "mid":                     "*.*",       # starts=-0.1, width=0.2
        "length":                  "0.2",       # e.g. 10 -> (-9.9, 10.1]
        "closed":                  "right"      # default left-open right-closed
    },                                          # for group structures
    "min_distance":             *.*,    # min distance between atoms
    "dim_larsen":               *,      # dimension in larsen algorithm, -1 if no neighboor if found
    "symmetry": {  # symprec
        # (str %.0e) int[1, 230]   "F m -3 m"         '
        "1e-1": {"number": 0, "symbol": "", "crystal_system": ""},
        "1e-2": {"number": 0, "symbol": "", "crystal_system": ""},
        "1e-5": {"number": 0, "symbol": "", "crystal_system": ""},
    },
    "cif":                      "*",     # cif string
    "poscar"                    "*",     # VASP POSCAR string

    "trajectory": {
        "nframes":                  2,          # int
        "cell":                     2x3x3,      # np.ndarray, angstrom
        "positions":                2x3x3,      # np.ndarray, angstrom
        "scaled_positions":         2x3x3,      # np.ndarray
        "forces":                   2x3x3,      # np.ndarray
        "volume":                   [*.*, *.*], # volume of each frame
        "enthalpy":                 [*.*, *.*], # enthalpy of each frame, np.nan if None, eV
        "enthalpy_per_atom":        [*.*, *.*], # enthalpy per atom of each frame, np.nan if None, eV
        "source_file":              [*.*, *.*], # source path of each frame
        "source_idx":               [*.*, *.*], # index in their source of each frame
        "source_dir":               "*",       # str, source dir
        "kabsch": {
            "kabsch_rot":           3x3 ndarray,     # rotation mat from 'r' to 'i' in kabsch algorithm
            "max_d_cell_i_r":       float,           # max abs delta between cell_r and cell_i
            "avg_d_cell_i_r":       float,           # mean abs delta between cell_r and cell_i
            "max_d_cell_roti_r":    float,           # max abs delta between rotated cell_i and cell_r
            "avg_d_cell_roti_r":    float,           # mean abs delta between rotated cell_i and cell_r
        }
        "shifted_d_frac":           22x3 ndarray,    # delta fractional coordinates 'r' minus 'i' shifted between [-0.5, 0.5]
        "strain": {
            "rot":                  3x3 ndarray,     # U of polar decomposition <r>=UP<i> 
            "strain":               3x3 ndarray,     # strain matrix, strain=P-I
            "avg_strain":           *.*,             # float, average strain
        }
    },

    "calyconfig": {
        "version": "",
        "icode": 1,
        ...
    },
    "dftconfig":                [<INCAR_3>],   # list of str，不要替换换行符
    "pseudopotential":          [<La>, <H>],   # list of str，与elements对应
    "donator": {"name": "", "email": ""},
    "deprecated":               False,  # bool
    "deprecated_reason":        "",     # str
    "last_updated_utc": datetime.utcnow(),

| Possible deprecated reason | comment|
| :-                         | :-     |
| `error enthalpy : optimization fail` | `enthalpy_per_atom` > 610612508(9) |
| `number of extracted structure <= <lte> in this prediction task` | Too less structures (`lte`) in one prediction |
| `error enthalpy : solitary and too small` | Solitary energy structure in each prediction (low energy side) |
| `error structure : isolated atoms`| atoms too discrate in crystal prediction |
| `error structure : unreasonable` |  |


CHANGELOG.md
## Unreleased (2023-06-12)

### New feature:

- **patch**: patch each property one-by-one([`a66cd53`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/a66cd53a3c1b9e0b0bc0362bb78afb33fd90ae04)) (by ixsluo)
- **patch**: add parallel patch([`d798190`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/d7981909a82e27916fe423dfc7d76632e2a2bc7b)) (by ixsluo)
- **patch**: add patch properties to records([`31ef0f4`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/31ef0f41537b587dbc7664340b728b375370bbd8)) (by ixsluo)
- **readout**: add readout to cdvae DataFrame([`35e63ec`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/35e63ec2deae5cc5fe095c94ff447739c96cc7ef)) (by ixsluo)
- **query**: merge query structure and trajectory together([`bfa6eda`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/bfa6eda2d1054b0035b583e867fe5fa5fa37b193)) (by ixsluo)
- **unique**: add UniqueFinder to add unique structures([`36e23ec`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/36e23ec065a4f7a9d1ba9081396790baca6dc23c)) (by ixsluo)
- **cleanup**: add clean deprecated records in uniqcol([`a473959`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/a473959b456e5faedb17dd3b617c62d4ba1e95bc)) (by ixsluo)
- **query**: add query all projection in query structures([`38a2f88`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/38a2f88ac7e6efc46a2336dbae90c6b79dac894b)) (by ixsluo)
- **itertools**: add standart pairwise, python at least 3.9([`b9918f6`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/b9918f6b7c075397c0d59fe028cd6e08f6d7a06f)) (by ixsluo)
- **cleanup**: deprecate large enthalpy and number of task([`d653bd2`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/d653bd2d8b7d32edd0570ac3a34cdb636864fa82)) (by ixsluo)
- **query**: change to class staticmethod([`590f0da`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/590f0da5efc0da412630efcb50e2460414ad8c22)) (by ixsluo)
- **query**: add query newer 'last_updated_utc'([`0b26f93`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/0b26f9317a1971a38cae8b1998d6b2b469958719)) (by ixsluo)
- **inputdat**: add vscenergy([`9a0ce23`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/9a0ce238c95cb7e4280b0ef52d812cc1aca6a315)) (by ixsluo)
- add extact ini opt and derectly insert to db([`2eb475b`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/2eb475bb5e68c32551afcc274c791b7dc42e3b13)) (by ixsluo)
- **prop**: add calculate properties of one frame([`81ebe00`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/81ebe00f41624a3f8ddc85e447a55890884176e7)) (by ixsluo)
- **query**: add query on calypso max index([`636c28f`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/636c28f97d9c48723f6faa0a98053a0e041e1cc8)) (by ixsluo)
- **login**: split login to db and get collection([`00a569c`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/00a569cc9e7dceb0887b7738175eb69c7be99a7f)) (by ixsluo)
- add extract ini and opt([`b3811b8`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/b3811b8946e75c2bc087922c61a62f0a89ee07dd)) (by ixsluo)
- **cleanup**: add error cleaning queryes([`8eebe90`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/8eebe90a9b387cb61450878c83ccfa2506d321c4)) (by ixsluo)
- **record**: add spacegroup to record([`a1f255a`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/a1f255a39f0eb6dffdbc47cf66731f88883e342f)) (by ixsluo)
- **unique**: add find unique structures([`12f3bc1`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/12f3bc126ea6f106b288bfa3cbf9496a07cb77e4)) (by ixsluo)
- **query**: properties are also returned for query structure and traj([`a7c8d46`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/a7c8d4611a620f5fb2419f10ee6c964b53654c34)) (by ixsluo)
- **utils**: move custom itertools to utils([`b53540d`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/b53540de085d9222a34e38ee36b73e10f70eeb25)) (by ixsluo)
- **query**: add query structure and group piplines([`00a15f6`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/00a15f6ee50f913dff0b74d70ac346a5bc02b240)) (by ixsluo)
- add rawrecord from ase Atoms classmethd(unfinished)([`ce3e8b6`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/ce3e8b6ab63e4d283a5045ddaed55660dd2cee59)) (by ixsluo)
- add maintain index([`92cfdf1`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/92cfdf18d268b2113cda596edad8e88614ae51d6)) (by ixsluo)
- add insert doc from data pkl([`fea1941`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/fea19413eb5bb06c7f5201f680655cac6d5e10e0)) (by ixsluo)
- add script pickle pandas to data([`8282420`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/828242030560f29c78ad4e5ea7b9369ea284ff79)) (by ixsluo)
- **calydb**: add numpy registory and timestamp([`d3a3a2a`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/d3a3a2a6bb4a8ec72561463eca5c606fb8c8aa23)) (by ixsluo)
- ensure material_id unique index([`8f753ea`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/8f753eaac077280268e593174ce5ed2a84e8f0b8)) (by ixsluo)
- add read_inputdat([`70a5730`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/70a57303be57e36df9c92161a0d210d8695703f2)) (by ixsluo)
- **calydb**: update default data dict([`f4d930f`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/f4d930faa2c03df06c469ce369a911467382405d)) (by ixsluo)
- **calydb**: add README of leagcy process for db([`5dd7e29`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/5dd7e29e073c4357d18ac76696c3fc1903db3a2f)) (by ixsluo)
- **calydb**: add binary serialization([`2b14cb4`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/2b14cb4e901765337cb6b9d8ea717c3d425f29b8)) (by ixsluo)
- **calydb**: add db connection and unittest on it([`3fd8c0f`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/3fd8c0f5cd4f893fd76b192a3eebb5210a5cb91b)) (by ixsluo)

### Bugs fixed:

- **property**: return -1 if CrystalNN failed in dim_larsen([`070b191`](https://github.com/ICCMS-CALYPSO/CALYPSO-kit/commit/070b19121db2550e8571e899c187165e26d29997)) (by ixsluo)

calypsokit/calydb/record.py
from collections import UserDict
from datetime import datetime

import numpy as np
from ase import Atoms
from ase.data import atomic_numbers, covalent_radii
from ase.spacegroup import get_spacegroup


class BaseRecordDict(UserDict):
    def update_time(self):
        self.data["last_updated_utc"] = datetime.utcnow()

    def todb(self):
        return self.data


class RecordDict(BaseRecordDict):
    def __init__(self, initdata=None):
        super().__init__(initdata)
        default_data = {
            "material_id":              "",     # str
            "source": {
                "name": "",             # calypso or materials project, etc.
                "index": 0,             # int index in this source, or None
            },
            "elements":                 [],     # list of str
            "nelements":                0,      # int
            "elemcount":                [],     # list of int, count of each element
            "species":                  [],     # list of str, species of each atom
            "formula":                  "",     # str, metal and alphabet order
            "reduced_formula":          "",     # str, metal and alphabet order
            "natoms":                   0,      # int
            "cell":                     np.zeros([3, 3]),  # np.ndarray, angstrom
            "positions":                np.zeros([0, 3]),  # np.ndarray, angstrom
            "scaled_positions":         np.zeros([0, 3]),  # np.ndarray
            "forces":                   np.zeros([0, 3]),  # np.ndarray
            "enthalpy":           0.0,
            "enthalpy_per_atom":  0.0,
            "volume":             0.0,
            "volume_per_atom":    0.0,
            "density":            0.0,
            "clospack_volume":          0.0,    # float, A^3
            "clospack_volume_per_atom": 0.0,    # float, A^3
            "clospack_density":         0.0,    # float, g/cm^3
            "pressure":                 0.0,    # float, GPa
            "pressure_range": {             # each pressure is set to a bin
                "mid":          "0.0",      # starts=-0.1, width=0.2
                "length":       "0.2",      # e.g. 10 -> (-9.9, 10.1]
                "closed":       "right"     # default left-open right-closed
            },                              # for group structures

            "trajectory": {
                "nframes":                  0,      # int
                "cell":                     np.zeros([0, 3, 3]),  # np.ndarray, angstrom
                "positions":                np.zeros([0, 0, 3]),  # np.ndarray, angstrom
                "scaled_positions":         np.zeros([0, 0, 3]),  # np.ndarray
                "forces":                   np.zeros([0, 0, 3]),  # np.ndarray
                "volume":                   [],     # np.ndarray, A^3
                "enthalpy":                 [],     # np.ndarray, eV
                "enthalpy_per_atom":        [],     # np.ndarray, eV
                "source_file":              [],     # source path of each frame
                "source_idx":               [],     # index in each source
                "source_dir":               "",     # source dir
            },

            "calyconfig": {
                "version": "",
                "icode": 0,
            },
            "dftconfig":                [],   # list of str，不要替换换行符
            "pseudopotential":          [],   # list of str，与elements对应
            "symmetry": {
                # symprec(str %.0e)
                #        int[1, 230] str ("F m -3 m")
                "1e-1": {"number": 0, "symbol": ""},
                "1e-2": {"number": 0, "symbol": ""},
                "1e-5": {"number": 0, "symbol": ""},
            },
            "donator": {"name": "", "email": ""},
            "deprecated":               False,  # bool
            "deprecated_reason":        "",     # str
            "last_updated_utc": datetime.utcnow(),
        }  # fmt: skip
        default_data.update(self.data)
        self.data = default_data
        self.check()

    def check(self):
        assert (
            len(self.data["elements"])
            == self.data["nelements"]
            == len(self.data["pseudopotential"])
        ), (
            'nelements not match '
            + f'{self.data["elements"]=} '
            + f'{self.data["nelements"]=} '
            + f'{self.data["pseudopotential"]=}'
        )
        assert (
            self.data["natoms"]
            == self.data["positions"].shape[0]
            == self.data["scaled_positions"].shape[0]
            == self.data["forces"].shape[0]
            == self.data["trajectory"]["positions"].shape[1]
            == self.data["trajectory"]["scaled_positions"].shape[1]
            == self.data["trajectory"]["forces"].shape[1]
        ), "natoms not match"
        assert (
            self.data["trajectory"]["nframes"]
            == self.data["trajectory"]["cell"].shape[0]
            == self.data["trajectory"]["positions"].shape[0]
            == self.data["trajectory"]["scaled_positions"].shape[0]
            == self.data["trajectory"]["forces"].shape[0]
            == len(self.data["trajectory"]["volume"])
            == len(self.data["trajectory"]["enthalpy"])
            == len(self.data["trajectory"]["enthalpy_per_atom"])
        ), "nframes not match"

    @classmethod
    def fromAtoms(cls, trajectory, **kwargs):
        formula = trajectory[-1].symbols.formula
        pressure = trajectory[-1].get("pressure", 0.0)
        pressure_range = cls.get_pressure_range(pressure)
        (
            density,
            volume,
            clospack_density,
            clospack_volume,
        ) = cls.get_density_clospack_density(trajectory[-1])
        symmetry = cls.wrapped_get_symmetry(trajectory[:-1])

        datadict = {
            "elements": list(formula.count().keys()),
            "nelements": len(formula.count()),
            "elemcount": list(formula.count().values()),
            "species": trajectory[-1].get_chemical_symbols(),
            "formula": formula.format("metal"),
            "reduced_formula": formula.reduce()[1].format("metal"),
            "natoms": len(trajectory[-1]),
            "cell": trajectory[-1].cell[:],
            "positions": trajectory[-1].positions,
            "scaled_positions": trajectory[-1].get_scaled_positions(),
            "forces": trajectory[-1].info.get(
                "forces", np.zeros([len(trajectory[-1]), 3]) * np.nan
            ),
            "enthalpy": trajectory[-1].info.get("enthalpy", np.nan),
            "enthalpy_per_atom": trajectory[-1].info.get("enthalpy_per_atom", np.nan),
            "volume": volume,
            "volume_per_atom": volume / len(trajectory[-1]),
            "density": density,
            "clospack_volume": clospack_volume,
            "clospack_volume_per_atom": clospack_volume / len(trajectory[-1]),
            "clospack_density": clospack_density,
            "pressure": trajectory[-1].get("pressure", 0.0),
            "pressure_range": pressure_range,
            "trajectory": {
                "nframes": len(trajectory),  # int
                "cell": np.stack([atoms.cell[:] for atoms in trajectory]),
                "positions": np.stack([atoms.positions for atoms in trajectory]),
                "scaled_positions": np.stack(
                    [atoms.get_scaled_positions() for atoms in trajectory]
                ),
                "forces": np.stack(
                    [
                        atoms.info.get("forces", np.zeros([len(atoms), 3]))
                        for atoms in trajectory
                    ]
                ),
                "volume": [atoms.get_volumes() for atoms in trajectory],
                "enthalpy": [
                    atoms.info.get("enthalpy", np.nan) for atoms in trajectory
                ],
                "enthalpy_per_atom": [
                    atoms.info.get("enthalpy_per_atom", np.nan) / len(atoms)
                    for atoms in trajectory
                ],
                "source": [atoms.info.get("source", None) for atoms in trajectory],
                "source_idx": [
                    atoms.info.get("source_idx", None) for atoms in trajectory
                ],
                "source_dir": trajectory[-1].info.get("source_dir", None),
            },
            "calyconfig": trajectory[-1].info.get("calyconfig", None),
            "dftconfig": [atoms.info.get("dftconfig", None) for atoms in trajectory],
            "pseudopotential": trajectory[-1].info.get(
                "pseudopotential", [None] * len(trajectory[-1])
            ),
            "symmetry": symmetry,
            # TODO: recommand to use multiprecessing to deal with core dump
            "donator": {"name": "", "email": ""},  # TODO: add donator
            "deprecated": False,
            "deprecated_reason": "",
        }
        return datadict

    @staticmethod
    def get_density_clospack_density(atoms: Atoms):
        mass = sum(atoms.get_masses())  # g/mol
        volume = atoms.get_volume()  # A^3
        density = mass / volume / 0.602214076  # to g/cm^3
        clospack_volume = sum(
            [
                4 / 3 * np.pi * covalent_radii[atomic_numbers[atom.symbol]] ** 3
                for atom in atoms
            ]
        )
        packingcoef = 0.74048
        clospack_volume /= packingcoef
        clospack_density = (
            mass / clospack_volume
        )  # (g/molA3)  * 6.02 * 1e-27 * 1e30(g/m3)
        return density, volume, clospack_density, clospack_volume

    @staticmethod
    def get_pressure_range(pressure, length=0.2, closed="right") -> dict:
        """put float pressure into string range

        Put the pressure to [-0.1, 0.1, 0.3, ...] bins centered on 0 (with length 0.2
        by default), return the mid of the bin as the representation. There is no worth
        setting length less than 0.1 and it is recommanded to keep 0.2 for compatibility
        with those already stored records.

        Because floating numbers cannot be strictly judged to be equal.
        This range is for group structures which are in a small pressure range.

        Parameters
        ----------
        pressure : float
            pressure
        length : float, optional
            bin length, , by default 0.2
        closed : str, optional
            right for (...] and left for [...), by default "right"

        Returns
        -------
        dict
            {"mid": "x.x", "length": "d.d", "closed": "<closed>"}, save as string format
            1 digit after the decimal point

        Raises
        ------
        NotImplementedError
            Argument closed is not "right" or "left"
        """
        length = abs(length)
        if closed == "right":
            mid = -((-pressure - length / 2) // length + 1) * length
        elif closed == "left":
            mid = ((pressure + length / 2) // length) * length
        else:
            raise NotImplementedError("Argument colsed only support [right|left]")
        if abs(mid) < 1e-3:
            mid = abs(mid)
        return {
            "mid": "{:.1f}".format(mid),
            "length": "{:.1f}".format(length),
            "closed": closed,
        }

    @staticmethod
    def get_symmetry(atoms: Atoms, symprec):
        # TODO: deprecated
        spg = get_spacegroup(atoms, symprec)
        return {"number": spg.no, "symbol": spg.symbol}

    def warpped_get_symmetry(self, atoms: Atoms):
        # TODO: deprecated
        return {
            key: self.get_symmetry(atoms, symprec)
            for key, symprec in zip(["1e-1", "1e-2", "1e-5"], [1e-1, 1e-2, 1e-5])
        }
